Unwrap `CompletionException`  in `StreamManagerImpl`

**Problem description**
The `StreamManager` interface that is exposed to app code provides synchronous methods, but deals with futures returned by the underlying `Controller`.  When an exception is thrown it should be unwrapped to remove the `CompletionException`, for the sake of app simplicity.

For example, here's a timeout exception from a call to `StreamManager::createScope`:
```
java.util.concurrent.CompletionException: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.grpc.StatusRuntimeException: UNAVAILABLE: NameResolver returned an empty list
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[na:1.8.0_151]
...
```
In the above case, it would simplify the app code to simply catch `RetriesExhaustedException`.   

**Problem location**


**Suggestions for an improvement**
Unwrap the exception or throw a more informative exception.

The current behavior of the apis under `StreamManagerImpl` is to wrap the exception with a `RuntimeException`.

```
Futures.getAndHandleExceptions(controller.createScope(scopeName), RuntimeException::new);
```
The exception thrown will be of the form (for `StreamManagerImpl#createScope`) is as shown below

```
java.lang.RuntimeException: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.grpc.StatusRuntimeException: UNAVAILABLE
	at io.pravega.common.concurrent.Futures.getAndHandleExceptions(Futures.java:181)
	at io.pravega.client.admin.impl.StreamManagerImpl.createScope(StreamManagerImpl.java:102)
	at io.pravega.test.integration.endtoendtest.EndToEndReaderGroupAgainstStandalone.testReader(EndToEndReaderGroupAgainstStandalone.java:77)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.grpc.StatusRuntimeException: UNAVAILABLE
	at io.pravega.common.util.Retry$RetryAndThrowBase.lambda$null$6(Retry.java:245)
	at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:870)
	at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:852)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
````

--- a/client/src/main/java/io/pravega/client/admin/impl/StreamManagerImpl.java
+++ b/client/src/main/java/io/pravega/client/admin/impl/StreamManagerImpl.java
@@ -58,69 +58,77 @@ public class StreamManagerImpl implements StreamManager {
 
     @Override
     public boolean createStream(String scopeName, String streamName, StreamConfiguration config) {
-        log.info("Creating scope/stream: {}/{} with configuration: {}", scopeName, streamName, config);
         NameUtils.validateUserStreamName(streamName);
-        return Futures.getAndHandleExceptions(controller.createStream(scopeName, streamName, StreamConfiguration.builder()
+        NameUtils.validateUserScopeName(scopeName);
+        log.info("Creating scope/stream: {}/{} with configuration: {}", scopeName, streamName, config);
+        return  Futures.getThrowingException(controller.createStream(scopeName, streamName, StreamConfiguration.builder()
                         .scalingPolicy(config.getScalingPolicy())
                         .retentionPolicy(config.getRetentionPolicy())
-                        .build()),
-                RuntimeException::new);
+                        .build()));
     }
 
     @Override
     public boolean updateStream(String scopeName, String streamName, StreamConfiguration config) {
+        NameUtils.validateUserStreamName(streamName);
+        NameUtils.validateUserScopeName(scopeName);
         log.info("Updating scope/stream: {}/{} with configuration: {}", scopeName, streamName, config);
-        return Futures.getAndHandleExceptions(controller.updateStream(scopeName, streamName, StreamConfiguration.builder()
+        return Futures.getThrowingException(controller.updateStream(scopeName, streamName, StreamConfiguration.builder()
                         .scalingPolicy(config.getScalingPolicy())
                         .retentionPolicy(config.getRetentionPolicy())
-                        .build()),
-                RuntimeException::new);
+                        .build()));
     }
 
     @Override
     public boolean truncateStream(String scopeName, String streamName, StreamCut streamCut) {
+        NameUtils.validateUserStreamName(streamName);
+        NameUtils.validateUserScopeName(scopeName);
         Preconditions.checkNotNull(streamCut);
-
         log.info("Truncating scope/stream: {}/{} with stream cut: {}", scopeName, streamName, streamCut);
-        return Futures.getAndHandleExceptions(controller.truncateStream(scopeName, streamName, streamCut),
-                RuntimeException::new);
+        return Futures.getThrowingException(controller.truncateStream(scopeName, streamName, streamCut));
     }
 
     @Override
     public boolean sealStream(String scopeName, String streamName) {
-        return Futures.getAndHandleExceptions(controller.sealStream(scopeName, streamName), RuntimeException::new);
+        NameUtils.validateUserStreamName(streamName);
+        NameUtils.validateUserScopeName(scopeName);
+        log.info("Sealing scope/stream: {}/{}", scopeName, streamName);
+        return Futures.getThrowingException(controller.sealStream(scopeName, streamName));
     }
 
     @Override
-    public boolean deleteStream(String scopeName, String toDelete) {
-        return Futures.getAndHandleExceptions(controller.deleteStream(scopeName, toDelete), RuntimeException::new);
+    public boolean deleteStream(String scopeName, String streamName) {
+        NameUtils.validateUserStreamName(streamName);
+        NameUtils.validateUserScopeName(scopeName);
+        log.info("Deleting scope/stream: {}/{}", scopeName, streamName);
+        return  Futures.getThrowingException(controller.deleteStream(scopeName, streamName));
     }
 
     @Override
     public boolean createScope(String scopeName) {
         NameUtils.validateUserScopeName(scopeName);
-        return Futures.getAndHandleExceptions(controller.createScope(scopeName),
-                RuntimeException::new);
-        
+        log.info("Creating scope: {}", scopeName);
+        return  Futures.getThrowingException(controller.createScope(scopeName));
     }
 
     @Override
     public boolean deleteScope(String scopeName) {
-        return Futures.getAndHandleExceptions(controller.deleteScope(scopeName),
-                RuntimeException::new);
+        NameUtils.validateUserScopeName(scopeName);
+        log.info("Deleting scope: {}", scopeName);
+        return  Futures.getThrowingException(controller.deleteScope(scopeName));
     }
 
     @Override
     public StreamInfo getStreamInfo(String scopeName, String streamName) {
+        NameUtils.validateUserStreamName(streamName);
+        NameUtils.validateUserScopeName(scopeName);
         log.info("Fetching StreamInfo for scope/stream: {}/{}", scopeName, streamName);
-        CompletableFuture<StreamInfo> streamInfo = getStreamInfo(Stream.of(scopeName, streamName));
-        return Futures.getAndHandleExceptions(streamInfo, RuntimeException::new);
+        return Futures.getThrowingException(getStreamInfo(Stream.of(scopeName, streamName)));
     }
 
     /**
      * Fetch the {@link StreamInfo} for a given stream.
-     * Note: The access level of this method can be reduced once the deprecated method {@link io.pravega.client.BatchClientFactory#getStreamInfo(Stream)}
-     * is removed.
+     * Note: The access level of this method can be reduced once the deprecated method
+     * {@link io.pravega.client.batch.BatchClient#getStreamInfo(Stream)} is removed.
      *
      * @param stream The Stream.
      * @return A future representing {@link StreamInfo}.

--- a/client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java
+++ b/client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java
@@ -976,7 +976,11 @@ public class ControllerImpl implements Controller {
         @Override
         public void onError(Throwable t) {
             log.warn("gRPC call for {} with trace id {} failed with server error.", method, traceId, t);
-            future.completeExceptionally(t);
+            if (t instanceof RuntimeException) {
+                future.completeExceptionally(t);
+            } else {
+                future.completeExceptionally(new RuntimeException(t));
+            }
         }
 
         @Override

